webapp/src/components/BidAssetPage/BidAssetForm.js

```diff
@@ -0,0 +1,184 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import addDays from 'date-fns/add_days'
+import dateFnsFormat from 'date-fns/format'
+import differenceInDays from 'date-fns/difference_in_days'
+import { Form, Button, Input, Message } from 'semantic-ui-react'
+import { t } from '@dapps/modules/translation/utils'
+
+import { assetType, bidType } from 'components/types'
+import { preventDefault, formatDate, formatMana } from 'lib/utils'
+import { isParcel } from 'shared/parcel'
+import {
+  DEFAULT_DAY_INTERVAL,
+  MINIMUM_DAY_INTERVAL,
+  MAXIMUM_BID_DAY_INTERVAL,
+  MINIMUM_ASSET_PRICE
+} from 'shared/listing'
+import TxStatus from 'components/TxStatus'
+
+const INPUT_FORMAT = 'YYYY-MM-DD'
+
+export default class BidAssetForm extends React.PureComponent {
+  static propTypes = {
+    asset: assetType,
+    assetName: PropTypes.string,
+    bid: bidType,
+    isTxIdle: PropTypes.bool,
+    isDisabled: PropTypes.bool,
+    onBid: PropTypes.func.isRequired,
+    onCancel: PropTypes.func.isRequired
+  }
+
+  constructor(props) {
+    super(props)
+    const { bid } = props
+
+    this.state = {
+      price: bid ? bid.price : '',
+      expiresAt: bid
+        ? dateFnsFormat(parseInt(bid.expires_at, 10), INPUT_FORMAT)
+        : this.formatFutureDate(DEFAULT_DAY_INTERVAL),
+      formErrors: []
+    }
+  }
+
+  handlePriceChange = e => {
+    // Dots and commas are not allowed
+    const price = e.currentTarget.value.replace(/\.|,/g, '')
```

Comment:
This is repeated in three files already, maybe we could abstract it?

Reply:
good Catch